java-checks/src/main/java/org/sonar/java/checks/regex/UnicodeAwareCharClassesCheck.java

```diff
@@ -0,0 +1,110 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.java.regex.RegexCheck;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.CharacterRangeTree;
+import org.sonar.java.regex.ast.EscapedCharacterClassTree;
+import org.sonar.java.regex.ast.NonCapturingGroupTree;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S5867")
+public class UnicodeAwareCharClassesCheck extends AbstractRegexCheck {
+
+  private static final Set<Character> unicodeAwareClassesWithFlag = ImmutableSet.of('s', 'S', 'w', 'W');
+  private static final Set<String> unicodeAwarePropertiesWithFlag = ImmutableSet.of(
+    "Lower", "Upper", "Alpha", "Alnum", "Punct", "Graph", "Print", "Blank", "Space");
+
+  private static final Map<Character, Character> unicodeUnawareCharacterRanges = ImmutableMap.of(
+    'a', 'z', 'A', 'Z');
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    UnicodeUnawareCharClassFinder unicodeUnawareCharClassFinder = new UnicodeUnawareCharClassFinder();
+    unicodeUnawareCharClassFinder.visit(regexForLiterals);
+
+    unicodeUnawareCharClassFinder.unicodeUnawareRange.forEach(unawareRange -> reportIssue(unawareRange,
+      "Replace this character range with a Unicode-aware character class.", null, Collections.emptyList()));
+
+    if (!unicodeUnawareCharClassFinder.unicodeAwareWithFlag.isEmpty() &&
+      !unicodeUnawareCharClassFinder.containsUnicodeCharacterFlag()) {
+
+      List<RegexCheck.RegexIssueLocation> secondaries = unicodeUnawareCharClassFinder.unicodeAwareWithFlag.stream()
+        .map(tree -> new RegexIssueLocation(tree, "Predefined/POSIX character class"))
+        .collect(Collectors.toList());
+      reportIssue(ExpressionUtils.methodName(mit),
+        "Enable the \"UNICODE_CHARACTER_CLASS\" flag or use a Unicode-aware alternative.", null, secondaries);
+    }
+  }
```

Comment:
I'd prefer to move the bulk of this method into `UnicodeUnawareCharClassFinder.after` (removing the `static` modifier from `UnicodeUnawareCharClassFinder`) and keep the tradition of `checkRegex` being implemented as only `new UnicodeUnawareCharClassFinder().visit(regexForLiterals);`.